src/module_library/secant_method.h

```diff
@@ -0,0 +1,180 @@
+ *
+ * \f[ x \maptso x - \frac{f(x)}{f'(x)} \f]
+ *
+ * to produce a sequence of approximations of the root. Given a good guess,
+ * the sequence will converge to the root rather quickly. The method fails
+ * when \f$ f'(x) = 0 \f$ and the sequences of approximate roots can
+ * diverge when close to a point where \f$ f'(x) = 0 \f$ nearly so.
+ *
+ * The method achieves its best performance when the root is a simple root:
+ * that is when the intersection of the graph of \f$f(x)\f$ with \f$y=0\f$
+ * is transversal and not tangent.
+ *
+ * For instance, Newton's method can identify the root of \f$ f(x) = x^2 \f$
+ * but the intersection is not transveral, \f$f'(x) = 0\f$ at the root.
+ * This pathology slows the method.
+ *
+ * The secant method uses the same logic as Newton's method but instead of
+ * using the first derivative, we approximate \f$f\f$ using a secant, computed
+ * from two points.
+ *
+ *  \f[ f(x) \approx f(x_1) + \left(\frac{f(x_1) - f(x_0)}{x_1 - x_0}\right) \left(x -x_1\right) \f]
+ *
+ * The root of this approximation is computed to produce to refine a guess for
+ * the root. Let \f$y_n = f(x_n)\f$; the secant method's update rule is:
+ *
+ * \f[ x_{n+1} = x_n - y_n \left(\frac{x_n - x_{n-1}}{y_n - y_{n-1}}\right)\f]
+ *
+ * This formula can be rearranged into
+ * Under ideal conditions, this method approaches the rate of convergence
+ * achieved in Newton's method. The same pathologies can also frustrate it.
+ *
+ * The secant method requires two starting guesses rather than one.
+ * Good starting guesses are important to achieving good performance. It is also
+ * important to identify potential pathologies.
+ *
+ * The order of initial guesses can affect convergence, so make the second guess
+ * the closer of the two if possible.
+ *
+ * @param [in] fun A function or object with `double operator()(double)` signature
+ *             This routine is templated so any type with a call operator with
+ *              the correct type signature will work.
+ * @param [in] x0 The first guess.
+ * @param [in] x1 The second guess.
+ *             type signature secant method calls this.
+ * @param [in] par The parameters for tolerances, maximum number of iterations, etc.
+ *              See the `secant_parameters` documentation. These are passed by
+ *             reference so the secant method routine can save the final function
+ *              evaluation (as a check on accuracy) and the number of iterations.
+ *
+ * @return Approximate root. If the method fails, it will return it's last (and
+ *      presumably best) guess for the root. If `check` member field in the
+ *      secant_parameters object is not zero, then the method has failed.
+ *      This routine will terminate if NaN or non-finite guesses are encountered.
+ */
+template <typename F>
+double find_root_secant_method(
+    F fun, double x0, double x1, secant_parameters& par)
+{
+    // check inputs for trivial solutions
```

Comment:
I don't think we have guidelines about this, but I feel that comments should start with a capital letter.